src/commands/login.rs

```diff
@@ -0,0 +1,424 @@
+use std::io::stdin;
+use std::path::PathBuf;
+use std::time::Duration;
+
+use anyhow::{bail, Context, Result};
+use clap::Parser;
+use cloud::client::{Client, ConnectionConfig};
+use cloud_openapi::models::DeviceCodeItem;
+use cloud_openapi::models::TokenInfo;
+use hippo::Client as HippoClient;
+use hippo::ConnectionInfo;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json::json;
+use tokio::fs;
+use tracing::log;
+use uuid::Uuid;
+
+use crate::opts::{
+    BINDLE_PASSWORD, BINDLE_SERVER_URL_OPT, BINDLE_URL_ENV, BINDLE_USERNAME, HIPPO_PASSWORD,
+    HIPPO_SERVER_URL_OPT, HIPPO_URL_ENV, HIPPO_USERNAME, INSECURE_OPT,
+};
+
+// this is the client ID registered in the Cloud's backend
+const SPIN_CLIENT_ID: &str = "583e63e9-461f-4fbe-a246-23e0fb1cad10";
+
+const DEFAULT_CLOUD_URL: &str = "http://localhost:5309";
+
+/// Log into the server
+#[derive(Parser, Debug)]
+#[clap(about = "Log into the server")]
+pub struct LoginCommand {
+    /// URL of bindle server
+    #[clap(
+        name = BINDLE_SERVER_URL_OPT,
+        long = "bindle-server",
+        env = BINDLE_URL_ENV,
+        requires = HIPPO_SERVER_URL_OPT,
+        requires = HIPPO_USERNAME,
+        requires = HIPPO_PASSWORD,
+    )]
+    pub bindle_server_url: Option<String>,
+
+    /// Basic http auth username for the bindle server
+    #[clap(
+        name = BINDLE_USERNAME,
+        long = "bindle-username",
+        env = BINDLE_USERNAME,
+        requires = BINDLE_PASSWORD
+    )]
+    pub bindle_username: Option<String>,
+
+    /// Basic http auth password for the bindle server
+    #[clap(
+        name = BINDLE_PASSWORD,
+        long = "bindle-password",
+        env = BINDLE_PASSWORD,
+        requires = BINDLE_USERNAME
+    )]
+    pub bindle_password: Option<String>,
+
+    /// Ignore server certificate errors from bindle and hippo
+    #[clap(
+        name = INSECURE_OPT,
+        short = 'k',
+        long = "insecure",
+        takes_value = false,
+    )]
+    pub insecure: bool,
+
+    /// URL of hippo server
+    #[clap(
+        name = HIPPO_SERVER_URL_OPT,
+        long = "hippo-server",
+        env = HIPPO_URL_ENV,
+        requires = BINDLE_SERVER_URL_OPT,
+        requires = HIPPO_USERNAME,
```

Comment:
At the moment if you pass a URL then you have to pass a username and password, even if you plan to use GH auth.

(sorry, thought I'd flagged this but apparently not)

Reply:
Good point. This is a holdover from prior iterations. Will change